flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/ProcessTableFunction.java

```diff
@@ -351,9 +439,158 @@ public interface Context {
+         * Thus, the returned timestamp represents the entire Flink subtask, independent of the
+         * currently processed partition. This behavior is similar to a call to {@code SELECT
+         * CURRENT_WATERMARK(...)} in SQL.
+         *
+         * <p>If a watermark was not received from all inputs, the method returns {@code null}.
+         *
+         * <p>In case this method is called within the {@code onTimer()} method, the returned
+         * watermark is the triggering watermark that currently fires the timer.
+         *
+         * @return the current watermark of the Flink subtask, or {@code null} if no common logical
+         *     time could be determined from the inputs
+         */
+        TimeType currentWatermark();
+
+        /**
+         * Registers a timer under the given name.
+         *
+         * <p>The timer fires when the {@link #currentWatermark()} advances the logical clock of the
+         * Flink subtask to a timestamp later or equal to the desired timestamp. In other words: A
+         * timer only fires if a watermark was received from all inputs and the timestamp is smaller
+         * or equal to the minimum of all received watermarks.
+         *
+         * <p>Timers can be named for distinguishing them in the {@code onTimer()} method.
+         * Registering a timer under the same name twice will replace an existing timer.
+         *
+         * <p>Note: Because only PTFs taking set semantic tables support state, and timers are a
+         * special kind of state, at least one {@link ArgumentTrait#TABLE_AS_SET} table argument
+         * must be declared.
+         *
+         * @param name identifier of the timer
+         * @param time timestamp when the timer should fire
+         */
+        void registerOnTime(String name, TimeType time);
+
+        /**
+         * Registers a timer.
+         *
+         * <p>The timer fires when the {@link #currentWatermark()} advances the logical clock of the
+         * Flink subtask to a timestamp later or equal to the desired timestamp. In other words: A
+         * timer only fires if a watermark was received from all inputs and the timestamp is smaller
+         * or equal to the minimum of all received watermarks.
+         *
+         * <p>Only one timer can be registered for a given time.
+         *
+         * <p>Note: Because only PTFs taking set semantic tables support state, and timers are a
+         * special kind of state, at least one {@link ArgumentTrait#TABLE_AS_SET} table argument
+         * must be declared.
+         *
+         * @param time timestamp when the timer should fire
+         */
+        void registerOnTime(TimeType time);
+
+        /**
+         * Clears a timer that was previously registered under the given name.
+         *
+         * <p>The call is ignored if no timer can be found.
+         *
+         * @param name identifier of the timer
+         */
+        void clearTimer(String name);
```

Comment:
Similar question about returning the value associated with `name` like in case of `Map#remove` for instance